File: jchucomponents-ktx/src/main/kotlin/com/jeluchu/jchucomponents/ktx/date/DateExtensions.kt
/*
 *
 *  Copyright 2022 Jeluchu
 *
 */

package com.jeluchu.jchucomponents.ktx.date

import java.text.DateFormatSymbols
import java.text.SimpleDateFormat
import java.util.*
import kotlin.math.absoluteValue

/**
 *
 * Constant to perform the date formatting as follows: "dd/MM/yyyy"
 * For example: "04/07/2022"
 *
 * d: Day
 * M: Month
 * y: Year
 *
 */
const val DATE_FORMAT_VERBOSE = "dd/MM/yyyy"

/**
 *
 * Constant to perform the date formatting as follows: "dd/MM/yyyy HH:mm"
 * For example: "04/07/2022 21:02"
 *
 * d: Day
 * M: Month
 * y: Year
 * H: Hour
 * m: Minutes
 *
 */
const val DATE_FORMAT_TIMESTAMP = "dd/MM/yyyy HH:mm"

/**
 *
 * Constant to perform the time formatting as follows: "HH:mm 'H'"
 * For example: "21:02 hrs"
 *
 * H: Hour
 * m: Minutes
 * 'H': Custom String param
 *
 */
const val DATE_FORMAT_ONLY_TIME = "HH:mm 'H'"

/**
 *
 * Constant to perform the time formatting as follows: "EEEE, MMMM d, yyyy - hh:mm:ss a"
 * For example: "21:02 hrs"
 *
 * E: Day of week
 * M: Month of year
 * d: Day of month
 * y: Year
 * H: Hour
 * m: Minutes
 * s: Seconds
 * a: AM / PM
 *
 */
const val DATE_FORMAT_WEEK_AND_MONTH_TIME = "EEEE, MMMM d, yyyy - hh:mm:ss a"

/** ---- FORMATS ------------------------------------------------------------------------------- **/

fun Date.format(): String =
    SimpleDateFormat(
        DATE_FORMAT_VERBOSE,
        Locale.getDefault()
    ).format(this)

fun Date.formatWithTime(): String =
    SimpleDateFormat(
        DATE_FORMAT_TIMESTAMP,
        Locale.getDefault()
    ).format(this)

fun Date.formatOnlyTime(): String =
    SimpleDateFormat(
        DATE_FORMAT_ONLY_TIME,
        Locale.getDefault()
    ).format(this)

fun simpleFormat(pattern: String) = SimpleDateFormat(pattern, Locale.getDefault())

/**
 * Pattern: yyyy-MM-dd HH:mm:ss
 */
fun Date.formatToServerDateTimeDefaults(): String = simpleFormat("yyyy-MM-dd HH:mm:ss").format(this)

/**
 * Pattern: yyyy-MM-dd
 */
fun Date.formatToServerDateDefaults(): String = simpleFormat("yyyy-MM-dd").format(this)


fun Date.formatToTruncatedDateTime(): String = simpleFormat("yyyyMMddHHmmss").format(this)

/**
 * Pattern: HH:mm:ss
 */
fun Date.formatToServerTimeDefaults(): String = simpleFormat("HH:mm:ss").format(this)

/**
 * Pattern: dd/MM/yyyy HH:mm:ss
 */
fun Date.formatToViewDateTimeDefaults(): String = simpleFormat("dd/MM/yyyy HH:mm:ss").format(this)

/**
 * Pattern: dd/MM/yyyy
 */
fun Date.formatToViewDateDefaults(): String {
    val sdf = SimpleDateFormat("dd/MM/yyyy", Locale.getDefault())
    return sdf.format(this)
}

/**
 * Pattern: HH:mm:ss
 */
fun Date.formatToViewTimeDefaults(): String {
    val sdf = SimpleDateFormat("HH:mm:ss", Locale.getDefault())
    return sdf.format(this)
}

/**
 * Add field date to current date
 */
fun Date.add(field: Int, amount: Int): Date {
    Calendar.getInstance().apply {
        time = this@add
        add(field, amount)
        return time
    }
}

/**
 * Convert time in **seconds** to `hh:mm:ss` or `mm:ss`
 */
fun Int.toDurationText(): String = this.run {
    if (this > 3600) "%d:%02d:%02d".format(this / 3600, (this % 3600) / 60, this % 60)
    else "%02d:%02d".format(this / 60, this % 60)
}

fun Date.addYears(years: Int): Date = add(Calendar.YEAR, years)
fun Date.addMonths(months: Int): Date = add(Calendar.MONTH, months)
fun Date.addDays(days: Int): Date = add(Calendar.DAY_OF_MONTH, days)
fun Date.addHours(hours: Int): Date = add(Calendar.HOUR_OF_DAY, hours)
fun Date.addMinutes(minutes: Int): Date = add(Calendar.MINUTE, minutes)
fun Date.addSeconds(seconds: Int): Date = add(Calendar.SECOND, seconds)

/** ---- CALENDAR ------------------------------------------------------------------------------ **/

fun Date.toCalendar(): Calendar {
    val calendar = Calendar.getInstance()
    calendar.time = this
    return calendar
}

fun Date.backInYears(years: Int): Date =
    Calendar.getInstance().also { calendar ->
        calendar.time = this
        calendar.add(Calendar.YEAR, -years)
    }.time

fun Date.forwardInYears(years: Int): Date =
    Calendar.getInstance().also { calendar ->
        calendar.time = this
        calendar.add(Calendar.YEAR, years)
    }.time

fun Date.isAfterOrEqualThan(numberDaysBeforeToday: Int, now: Date? = null): Boolean {
    val calendar = Calendar.getInstance()
    now?.let { calendar.time = it }
    calendar[Calendar.HOUR_OF_DAY] = 0
    calendar[Calendar.MINUTE] = 0
    calendar[Calendar.SECOND] = 0
    calendar[Calendar.MILLISECOND] = 0
    calendar.add(Calendar.DAY_OF_YEAR, -numberDaysBeforeToday)
    val backDate = calendar.time
    return this.time >= backDate.time
}

fun Date.minusDays(days: Int) = plusDays(-days)

fun Date.plusDays(days: Int): Date {
    val calendar = Calendar.getInstance()
    calendar.time = this
    calendar.add(Calendar.DATE, days)
    return calendar.time
}

fun Date.isBeforeThan(numberDaysBeforeToday: Int, now: Date? = null) =
    !isAfterOrEqualThan(numberDaysBeforeToday, now)

fun Date.getDateTime(daysNumber: Int): Date =
    Calendar.getInstance()
        .also {
            it.time = this
            it.add(Calendar.DAY_OF_MONTH, -daysNumber)
        }
        .time

fun Date.getDayEndCalendar(): Calendar =
    Calendar.getInstance().apply {
        time = this@getDayEndCalendar
        set(Calendar.HOUR_OF_DAY, 23)
        set(Calendar.MINUTE, 59)
        set(Calendar.SECOND, 59)
        set(Calendar.MILLISECOND, 999)
    }

fun Date.firstDayOfTheMonth(): Date {
    val calendar = Calendar.getInstance()
    calendar.time = firstHourOfTheDay()
    calendar.set(Calendar.DAY_OF_MONTH, calendar.getActualMinimum(Calendar.DAY_OF_MONTH))
    return calendar.time
}

fun Date.firstHourOfTheDay(): Date {
    val calendar = Calendar.getInstance()
    calendar.time = this
    calendar.set(Calendar.HOUR_OF_DAY, 0) // ! clear would not reset the hour of day !
    calendar.clear(Calendar.MINUTE)
    calendar.clear(Calendar.SECOND)
    calendar.clear(Calendar.MILLISECOND)
    return calendar.time
}

fun Date.plusMinutes(minutes: Int): Date {
    val calendar = Calendar.getInstance()
    calendar.time = this
    calendar.add(Calendar.MINUTE, minutes)
    return calendar.time
}

fun Date.diffInDays(next: Date): Long {
    val diffInMillis = next.time - time
    return (diffInMillis / (24 * 60 * 60 * 1000)).absoluteValue
}

fun Date.toAccessibilityDateMMMMYYYY(): CharSequence {
    val cal = Calendar.getInstance()
    cal.time = this
    return DateFormatSymbols().months[cal.get(Calendar.MONTH)] + " " + cal.get(Calendar.YEAR)
}

File: swift/Sources/JchuComponentsExtensions/Foundation/DateExtensions.swift
import Foundation

/// Day, month and year display pattern.
public let jchuDateFormatVerbose = "dd/MM/yyyy"
/// Day, month, year, hour and minute display pattern.
public let jchuDateFormatTimestamp = "dd/MM/yyyy HH:mm"
/// Hour and minute display pattern.
public let jchuDateFormatOnlyTime = "HH:mm 'H'"
/// Localized weekday, month and time display pattern.
public let jchuDateFormatWeekAndMonthTime = "EEEE, MMMM d, yyyy - hh:mm:ss a"

public extension DateFormatter {
    /// Creates a formatter with an explicit pattern, locale and time zone.
    static func jchu(
        pattern: String,
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> DateFormatter {
        let formatter = DateFormatter()
        formatter.dateFormat = pattern
        formatter.locale = locale
        formatter.timeZone = timeZone
        return formatter
    }
}

public extension Date {
    /// Formats this date as day, month and year.
    func format(
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> String {
        formatted(pattern: jchuDateFormatVerbose, locale: locale, timeZone: timeZone)
    }

    /// Formats this date as day, month, year, hour and minute.
    func formatWithTime(
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> String {
        formatted(pattern: jchuDateFormatTimestamp, locale: locale, timeZone: timeZone)
    }

    /// Formats this date as hour and minute.
    func formatOnlyTime(
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> String {
        formatted(pattern: jchuDateFormatOnlyTime, locale: locale, timeZone: timeZone)
    }

    /// Formats this date with an explicit Unicode date pattern.
    func formatted(
        pattern: String,
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> String {
        DateFormatter.jchu(
            pattern: pattern,
            locale: locale,
            timeZone: timeZone
        ).string(from: self)
    }

    func formatToServerDateTimeDefaults(
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> String {
        formatted(pattern: "yyyy-MM-dd HH:mm:ss", locale: locale, timeZone: timeZone)
    }

    func formatToServerDateDefaults(
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> String {
        formatted(pattern: "yyyy-MM-dd", locale: locale, timeZone: timeZone)
    }

    func formatToTruncatedDateTime(
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> String {
        formatted(pattern: "yyyyMMddHHmmss", locale: locale, timeZone: timeZone)
    }

    func formatToServerTimeDefaults(
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> String {
        formatted(pattern: "HH:mm:ss", locale: locale, timeZone: timeZone)
    }

    func formatToViewDateTimeDefaults(
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> String {
        formatted(pattern: "dd/MM/yyyy HH:mm:ss", locale: locale, timeZone: timeZone)
    }

    func formatToViewDateDefaults(
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> String {
        formatted(pattern: "dd/MM/yyyy", locale: locale, timeZone: timeZone)
    }

    func formatToViewTimeDefaults(
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> String {
        formatted(pattern: "HH:mm:ss", locale: locale, timeZone: timeZone)
    }

    /// Adds a calendar component, returning this date if calculation fails.
    func adding(
        _ component: Calendar.Component,
        amount: Int,
        calendar: Calendar = .current
    ) -> Date {
        calendar.date(byAdding: component, value: amount, to: self) ?? self
    }

    func addYears(_ years: Int, calendar: Calendar = .current) -> Date {
        adding(.year, amount: years, calendar: calendar)
    }

    func addMonths(_ months: Int, calendar: Calendar = .current) -> Date {
        adding(.month, amount: months, calendar: calendar)
    }

    func addDays(_ days: Int, calendar: Calendar = .current) -> Date {
        adding(.day, amount: days, calendar: calendar)
    }

    func addHours(_ hours: Int, calendar: Calendar = .current) -> Date {
        adding(.hour, amount: hours, calendar: calendar)
    }

    func addMinutes(_ minutes: Int, calendar: Calendar = .current) -> Date {
        adding(.minute, amount: minutes, calendar: calendar)
    }

    func addSeconds(_ seconds: Int, calendar: Calendar = .current) -> Date {
        adding(.second, amount: seconds, calendar: calendar)
    }

    func minusDays(_ days: Int, calendar: Calendar = .current) -> Date {
        addDays(-days, calendar: calendar)
    }

    func plusDays(_ days: Int, calendar: Calendar = .current) -> Date {
        addDays(days, calendar: calendar)
    }

    func plusMinutes(_ minutes: Int, calendar: Calendar = .current) -> Date {
        addMinutes(minutes, calendar: calendar)
    }

    func backInYears(_ years: Int, calendar: Calendar = .current) -> Date {
        addYears(-years, calendar: calendar)
    }

    func forwardInYears(_ years: Int, calendar: Calendar = .current) -> Date {
        addYears(years, calendar: calendar)
    }

    func firstHourOfTheDay(calendar: Calendar = .current) -> Date {
        calendar.startOfDay(for: self)
    }

    func firstDayOfTheMonth(calendar: Calendar = .current) -> Date {
        let components = calendar.dateComponents([.year, .month], from: self)
        return calendar.date(from: components) ?? firstHourOfTheDay(calendar: calendar)
    }

    /// Returns the final representable instant of this date's calendar day.
    func dayEnd(calendar: Calendar = .current) -> Date {
        let start = firstHourOfTheDay(calendar: calendar)
        return calendar.date(
            byAdding: DateComponents(day: 1, second: -1, nanosecond: 999_000_000),
            to: start
        ) ?? self
    }

    func isAfterOrEqualThan(
        numberDaysBeforeToday: Int,
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> Bool {
        let startOfToday = calendar.startOfDay(for: now)
        let backDate = calendar.date(
            byAdding: .day,
            value: -numberDaysBeforeToday,
            to: startOfToday
        ) ?? startOfToday

        return self >= backDate
    }

    func isBeforeThan(
        numberDaysBeforeToday: Int,
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> Bool {
        !isAfterOrEqualThan(
            numberDaysBeforeToday: numberDaysBeforeToday,
            now: now,
            calendar: calendar
        )
    }

    func getDateTime(daysNumber: Int, calendar: Calendar = .current) -> Date {
        addDays(-daysNumber, calendar: calendar)
    }

    /// Returns the absolute number of calendar-day boundaries between dates.
    func diffInDays(to next: Date, calendar: Calendar = .current) -> Int {
        let start = calendar.startOfDay(for: self)
        let end = calendar.startOfDay(for: next)
        return abs(calendar.dateComponents([.day], from: start, to: end).day ?? 0)
    }

    func toAccessibilityDateMMMMYYYY(
        locale: Locale = .current,
        timeZone: TimeZone = .current
    ) -> String {
        formatted(pattern: "MMMM yyyy", locale: locale, timeZone: timeZone)
    }
}

public extension Int {
    /// Formats a duration expressed in seconds as `mm:ss` or `h:mm:ss`.
    var durationText: String {
        if self >= 3600 {
            return String(format: "%d:%02d:%02d", self / 3600, (self % 3600) / 60, self % 60)
        }

        return String(format: "%02d:%02d", self / 60, self % 60)
    }
}

/// Returns the current date.
public func now() -> Date {
    Date()
}

public func currentDayWeekString(calendar: Calendar = .current) -> Int {
    calendar.component(.weekday, from: Date())
}

public func currentDayMonthString(calendar: Calendar = .current) -> Int {
    calendar.component(.day, from: Date())
}

public func currentMonthString(calendar: Calendar = .current) -> Int {
    calendar.component(.month, from: Date()) - 1
}

public func currentYearString(calendar: Calendar = .current) -> String {
    String(calendar.component(.year, from: Date()))
}

/// Formats a Unix timestamp expressed in milliseconds as a display date.
public func getDateTime(
    currentMillis: String,
    locale: Locale = .current,
    timeZone: TimeZone = .current
) -> String? {
    guard let milliseconds = TimeInterval(currentMillis) else {
        return nil
    }

    return Date(timeIntervalSince1970: milliseconds / 1000).formatted(
        pattern: "dd MMM yyyy",
        locale: locale,
        timeZone: timeZone
    )
}

/// Whether a millisecond timestamp belongs to a different calendar day.
public func isNextDay(
    lastFetchTime: TimeInterval,
    now: Date = Date(),
    calendar: Calendar = .current
) -> Bool {
    !calendar.isDate(Date(timeIntervalSince1970: lastFetchTime / 1000), inSameDayAs: now)
}

/// Whether a duration in milliseconds has elapsed since a millisecond timestamp.
public func isCustomTimePassed(
    lastFetchTime: TimeInterval,
    time: TimeInterval = 480 * 60 * 60 * 1000,
    now: Date = Date()
) -> Bool {
    now.timeIntervalSince1970 * 1000 - lastFetchTime >= time
}

public func isFetchFiveMinutes(lastFetchTime: TimeInterval, now: Date = Date()) -> Bool {
    isCustomTimePassed(lastFetchTime: lastFetchTime, time: 5 * 60 * 1000, now: now)
}

public func isFetchThirtyMinutes(lastFetchTime: TimeInterval, now: Date = Date()) -> Bool {
    isCustomTimePassed(lastFetchTime: lastFetchTime, time: 30 * 60 * 1000, now: now)
}

public func isFetchSixHours(lastFetchTime: TimeInterval, now: Date = Date()) -> Bool {
    isCustomTimePassed(lastFetchTime: lastFetchTime, time: 6 * 60 * 60 * 1000, now: now)
}

public func isFetchTwelveHours(lastFetchTime: TimeInterval, now: Date = Date()) -> Bool {
    isCustomTimePassed(lastFetchTime: lastFetchTime, time: 12 * 60 * 60 * 1000, now: now)
}

public func isFetchOneDay(lastFetchTime: TimeInterval, now: Date = Date()) -> Bool {
    isCustomTimePassed(lastFetchTime: lastFetchTime, time: 24 * 60 * 60 * 1000, now: now)
}

public func isFetchTwoDays(lastFetchTime: TimeInterval, now: Date = Date()) -> Bool {
    isCustomTimePassed(lastFetchTime: lastFetchTime, time: 48 * 60 * 60 * 1000, now: now)
}

public func isFetchThreeDays(lastFetchTime: TimeInterval, now: Date = Date()) -> Bool {
    isCustomTimePassed(lastFetchTime: lastFetchTime, time: 72 * 60 * 60 * 1000, now: now)
}

public func isFetchFourDays(lastFetchTime: TimeInterval, now: Date = Date()) -> Bool {
    isCustomTimePassed(lastFetchTime: lastFetchTime, time: 96 * 60 * 60 * 1000, now: now)
}

public func isFetchFiveDays(lastFetchTime: TimeInterval, now: Date = Date()) -> Bool {
    isCustomTimePassed(lastFetchTime: lastFetchTime, time: 120 * 60 * 60 * 1000, now: now)
}

public func isFetchSixDays(lastFetchTime: TimeInterval, now: Date = Date()) -> Bool {
    isCustomTimePassed(lastFetchTime: lastFetchTime, time: 144 * 60 * 60 * 1000, now: now)
}

public func isFetchSevenDays(lastFetchTime: TimeInterval, now: Date = Date()) -> Bool {
    isCustomTimePassed(lastFetchTime: lastFetchTime, time: 168 * 60 * 60 * 1000, now: now)
}

public func isFetchTenDays(lastFetchTime: TimeInterval, now: Date = Date()) -> Bool {
    isCustomTimePassed(lastFetchTime: lastFetchTime, time: 240 * 60 * 60 * 1000, now: now)
}

public func isFetchFifteenDays(lastFetchTime: TimeInterval, now: Date = Date()) -> Bool {
    isCustomTimePassed(lastFetchTime: lastFetchTime, time: 360 * 60 * 60 * 1000, now: now)
}

public func isFetchTwentyDays(lastFetchTime: TimeInterval, now: Date = Date()) -> Bool {
    isCustomTimePassed(lastFetchTime: lastFetchTime, time: 480 * 60 * 60 * 1000, now: now)
}
